modules/internal/String.chpl

```diff
@@ -17,572 +17,1436 @@
-  
-  inline proc _cast(type t, x:c_string) where t == real(32)
-    return c_string_to_real32(x);
-  inline proc _cast(type t, x:c_string) where t == real(64)
-    return c_string_to_real64(x);
-  inline proc _cast(type t, x:c_string) where t == imag(32)
-    return c_string_to_imag32(x);
-  inline proc _cast(type t, x:c_string) where t == imag(64)
-    return c_string_to_imag64(x);
+    // Make this some sort of local check once we have local types/vars
+    if (rhs_c.locale.id != chpl_nodeID) || (lhs.locale.id != chpl_nodeID) then
+      halt("Cannot assign a remote c_string to a string.");
 
-  //
-  // casts from c_string to complex types
-  //
-  pragma "insert line file info"
-  extern proc c_string_to_complex64(x:c_string) : complex(64);
-  pragma "insert line file info"
-  extern proc c_string_to_complex128(x:c_string) : complex(128);
+    const len = rhs_c.length;
+    const buff:bufferType = rhs_c:bufferType;
+    lhs.reinitString(buff, len, len+1, needToCopy=true);
 
-  inline proc _cast(type t, x:c_string) where t == complex(64)
-    return c_string_to_complex64(x);
-  inline proc _cast(type t, x:c_string) where t == complex(128)
-    return c_string_to_complex128(x);
+    if debugStrings then
+      chpl_debug_string_print("leaving proc =() c_string");
+  }
 
   //
-  // casts from complex
+  // Concatenation
   //
-  inline proc _cast(type t, x: complex(?w)) where t == c_string_copy {
-    if isnan(x.re) || isnan(x.im) then
-      return __primitive("string_copy", "nan");
-    var re = (x.re):c_string_copy;
-    var im: c_string_copy;
-    var op: c_string;
-    if x.im < 0 {
-      im = (-x.im):c_string_copy;
-      op = " - ";
-    } else if im == "-0.0" {
-      im = "0.0":c_string_copy;
-      op = " - ";
+  proc +(s0: string, s1: string) {
+    if debugStrings then
+      chpl_debug_string_print("in proc +()");
+
+    // cache lengths locally
+    const s0len = s0.len;
+    if s0len == 0 then return s1;
+    const s1len = s1.len;
+    if s1len == 0 then return s0;
+
+    var ret: string;
+    ret.len = s0len + s1len;
+    const allocSize = chpl_mem_goodAllocSize((ret.len+1).safeCast(size_t));
+    ret._size = allocSize.safeCast(int);
+    ret.buff = chpl_mem_alloc(allocSize,
+                              CHPL_RT_MD_STR_COPY_DATA): bufferType;
+    ret.owned = true;
+
+    const s0remote = s0.locale.id != chpl_nodeID;
+    if s0remote {
+      chpl_string_comm_get(ret.buff, s0.locale.id,
+                           s0.buff, s0len.safeCast(size_t));
     } else {
-      im = (x.im):c_string_copy;
-      op = " + ";
-    }
-    // TODO: Add versions of the concatenation operator that consume their
-    // c_string_copy arg or args.
-    var ts0 = re + op;
-    chpl_free_c_string_copy(re);
-    var ts1 = ts0 + im;
-    chpl_free_c_string_copy(ts0);
-    chpl_free_c_string_copy(im);
-    const ret = ts1 + "i";
-    chpl_free_c_string_copy(ts1);
+      memmove(ret.buff, s0.buff, s0len.safeCast(size_t));
```

Comment:
please use memcpy here. You just allocated ret.buff so it can't possibly alias s0.buff.